### ISSUE TITLE
bug: P800 FP16 F.linear 在大输出维度返回错误 logits（Qwen3.6-27B lm-head）

### ISSUE THREAD

**Issue by AlanJager** [baidu/vLLM-Kunlun#387]
### Bug Description

在昆仑 P800 上运行 Qwen3.6-27B Dense 时，模型可以成功加载且 `/health=200`，但 FP16 文本生成会持续选择错误 token，表现为重复、不可读文本。

沿 `compute_logits()` 边界对同一份 final hidden 和同一份运行时权重做数值差分后，问题可以缩小到以下公开调用边界：

```python
torch.nn.functional.linear(hidden_states, lm_head.weight)
```

对应 GEMM shape 为：

```text
M=1, N=248320, K=5120
hidden:         [1, 5120], FP16
lm_head.weight: [248320, 5120], FP16, contiguous
logits:         [1, 248320]
```

在当前环境中，单次完整 XPU `F.linear` 会产生严重错误 logits；将完全相同的 XPU hidden 和 weight 仅沿 vocab 行分块计算，再 `torch.cat`，结果会恢复并与 CPU FP32 reference 对齐。因此该问题不依赖 vLLM 服务即可独立复现。

#### 环境

| 项目 | 实测值 |
| --- | --- |
| XPU | 单张 P800 PCIe 96 GiB，TP1 |
| OS / kernel | Ubuntu 22.04，Linux `5.15.0-161-generic` |
| Driver / XPU-RT | `5.0.21.26` / `5.0.21` |
| 镜像 tag | `vllm-kunlun:p800-aios-xspeedgate-1.1.0-cocopod-1.1.0` |
| 本地 image ID | `sha256:c9935b8959c7c78c49e76ae397006534178c9451199cf220975ed16a146c2ae3` |
| vLLM / vLLM-Kunlun | `0.15.1` / `0.15.1.dev0` |
| vLLM-Kunlun commit | `4885de2166244a0ba1e76bab8b2a0537c20d3379` |
| Torch / Transformers / Triton | `2.5.1+cu118`（定制 XPU 版本）/ `5.2.0` / `3.1.0` |
| `kunlun_ops` | `0.1.58+ee39020a` |
| `xspeedgate_ops` | `1.1.0+53992ca` |
| `cocopod` | `1.1.0` |
| 模型 | Qwen3.6-27B，revision `6a9e13bd6fc8f0983b9b99948120bc37f49c13e9` |
| lm-head | unquantized、无 bias、运行时 FP16，`[248320,5120]` |

模型的 config、weight index、tokenizer 以及 15/15 个 safetensors 分片均已与上述官方 revision 逐项校验 SHA256。运行时完整 lm-head 共 `1,271,398,400` 个元素，与官方 BF16 权重 cast FP16 后逐元素比较为 `mismatch_count=0, max_abs=0`。

### Steps to Reproduce

下面的最小脚本只依赖定制 Torch、`safetensors` 和模型权重，不启动 vLLM。它执行以下对照：

1. 从模型分片读取真实 `lm_head.weight`，cast 为 contiguous FP16；
2. 使用固定 seed 生成 contiguous FP16 hidden `[1,5120]`；
3. CPU 使用 FP32 累加、每 4096 行计算 reference；
4. XPU 使用 FP16，并改变单次 `F.linear` 的最大 weight 行数；
5. 拼回完整 `[1,248320]` logits，比较全部元素、NaN/Inf、relative L2 和 top-1。

```python
import json
import math
import sys
from pathlib import Path

import torch
import torch.nn.functional as F
from safetensors import safe_open


model_dir = Path(sys.argv[1])
weight_name = "lm_head.weight"

with (model_dir / "model.safetensors.index.json").open() as f:
    index = json.load(f)
shard = model_dir / index["weight_map"][weight_name]

with safe_open(shard, framework="pt", device="cpu") as f:
    weight_cpu = f.get_tensor(weight_name).to(torch.float16).contiguous()


def cpu_reference(hidden, weight, rows=4096):
    parts = []
    for start in range(0, weight.shape[0], rows):
        end = min(start + rows, weight.shape[0])
        parts.append(F.linear(hidden.float(), weight[start:end].float()))
    return torch.cat(parts, dim=-1)


def xpu_projection(hidden, weight, rows):
    parts = []
    for start in range(0, weight.shape[0], rows):
        end = min(start + rows, weight.shape[0])
        parts.append(F.linear(hidden, weight[start:end]))
    return torch.cat(parts, dim=-1)


weight_xpu = weight_cpu.to("xpu")
for seed in (20260714, 20260715, 20260716):
    generator = torch.Generator(device="cpu").manual_seed(seed)
    hidden_cpu = torch.randn(
        (1, weight_cpu.shape[1]), generator=generator, dtype=torch.float32
    ).to(torch.float16).contiguous()
    reference = cpu_reference(hidden_cpu, weight_cpu)
    hidden_xpu = hidden_cpu.to("xpu")

    for rows in (32768, 65536, 81920, 98304, 99328, 100352, 248320):
        output = xpu_projection(hidden_xpu, weight_xpu, rows).float().cpu()
        finite = torch.isfinite(output)
        finite_all = bool(finite.all().item())
        relative_l2 = (
            float(torch.linalg.vector_norm(output - reference).item()
                  / torch.linalg.vector_norm(reference).item())
            if finite_all else math.inf
        )
        print({
            "seed": seed,
            "max_rows_per_gemm": rows,
            "finite": int(finite.sum().item()),
            "numel": output.numel(),
            "nan": int(torch.isnan(output).sum().item()),
            "inf": int(torch.isinf(output).sum().item()),
            "relative_l2": relative_l2,
            "top1": int(output.argmax(dim=-1).item()),
            "reference_top1": int(reference.argmax(dim=-1).item()),
        })
```

运行：

```bash
python reproduce_xpu_lm_head_gemm.py /path/to/Qwen3.6-27B
```

#### 三组 seed 的实测结果（K=5120）

表中的行数是每次 GEMM 的最大 `N`；各块最终按原顺序拼回完整 248320 个 logits。

| 单次最大输出行数 | 实测结果 |
| ---: | --- |
| `32768/65536/81920/98304` | 3/3 全部 finite，top-1 与 CPU 相同，relative L2 `3.011e-4..3.060e-4` |
| `99328` | 3/3 全部 finite、top-1 碰巧相同，但 relative L2 已达到 `0.1288..0.1524` |
| `100352` | relative L2 `0.1843..0.2110`，其中一组 top-1 错误 |
| `248320` | 三次均固定出现 1178 NaN、2 Inf，三组 top-1 均错误 |

`99328` 的结果说明，仅检查 finite 或一次 top-1 相同会漏掉严重的静默数值失真。

#### 真实模型请求的因果 A/B

固定：

```text
prompt      = "The capital of France is"
token IDs   = [760, 6511, 314, 9338, 369]
temperature = 0
max_tokens  = 1
TP          = 1
dtype       = float16
```

在 `compute_logits` 边界截取同一个 final hidden 和同一份运行时权重：

| 指标 | 原生 XPU 单次完整 GEMM | CPU FP32 reference |
| --- | --- | --- |
| top-1 | 错误，396 个 token 并列最大 | token `11751`，`" Paris"` |
| top-1 value | `32.0` | `15.637904` |
| top-1 margin | `0` | `2.847091` |
| cosine | `0.2263683` | `1.0`（reference） |
| relative L2 | `1.4691908` | `0`（reference） |
| max abs | `39.1820335` | `0`（reference） |

只把 lm-head 的计算替换为 CPU reference，API 首 token 立即恢复为 `" Paris"`；改成 4096 或 65536 行的多次 XPU `F.linear` 后也恢复为 `" Paris"`，且全部 logits finite、top-1 与 CPU 相同。

### Expected Behavior

对于相同的 contiguous FP16 输入和权重，P800 XPU `F.linear` 应返回有限且满足正常 FP16 误差范围的 logits。改变 `N` 或分块方式不应导致：

- 静默出现 relative L2 `0.1+` 的失真；
- NaN/Inf；
- top-1 token 改变；
- 模型持续输出错误或重复 token。

#### 已验证事实与尚未确认的内部根因

当前能够确认的是：错误发生在定制 XPU Torch 的大输出维度 FP16 `F.linear` 公开调用边界。结果会随 `(M,N,K,dtype)` 和分块 shape 离散变化。

我们推测不同 shape 选择了不同 GEMM kernel、tiling 或 workspace 路径，但这些实现位于闭源二进制中，目前无法确认具体内部原因。因此本 Issue 不把某个具体 kernel 或 tiling 参数声明为已定位根因。

`98304` 只是当前硬件、wheel、shape 和三组 seed 的实测对齐点，不是通用安全阈值。

#### K=2048 的补充 shape 证据

另用 Qwen3.5-35B-A3B 的真实 BF16 lm-head 权重 cast FP16，测试 `[248320,2048]`：

| 单次最大输出行数 | 结果 |
| ---: | --- |
| `32768/65536/81920/98304` | 三组 seed 全部 finite，top-1 与 CPU 相同，relative L2 `2.069e-4..2.078e-4` |
| `100352` | 两个补充 seed 均固定出现 58 个 NaN，top-1 错误 |
| `248320` | 三组 seed 均固定出现 2316 个 NaN、6 个 Inf，top-1 错误 |

这组数据只用于证明 `K=2048` 也存在 shape 相关的 GEMM 数值错误，不等价于 Qwen3.6 MoE 端到端验证。

#### 当前诊断 workaround 的边界

将 TP1 lm-head 沿 vocab 行拆成 4096 或 65536 行的多次 XPU `F.linear`，再在 XPU 上 `torch.cat`，可以恢复 CPU reference 和短文本生成；在线 workaround 没有使用 CPU。

该 workaround 仅用于证明“只改变 GEMM shape 即恢复正确性”，不是生产修复：

- 只验证 TP1、FP16、unquantized、无 bias；
- TP2 vocab shard、padding 和 gather 尚未纳入该补丁；
- 未完成无诊断插桩的性能、长上下文、并发和稳定性测试；
- 真实修复仍需要匹配的定制 Torch/XMLIR/XDNN 二进制。

#### 希望维护者协助

1. 使用匹配的闭源 wheel 运行上述最小复现，确认 shape 矩阵是否可以复现。
2. 确认这些 `F.linear` shape 实际选择的 GEMM kernel、tiling、workspace 及支持合同。
3. 提供可记录或强制 GEMM 算法选择的调试开关，帮助进一步缩小内部路径。
4. 提供修复后的 Torch/XMLIR/XDNN wheel，以及完整兼容版本矩阵。
5. 将 `M=1, K=2048/5120, N=98304/99328/100352/124160/248320` 加入完整向量数值回归；不能只检查 finite 或 top-1。
6. 如果短期无法替换底层二进制，请建议 vLLM-Kunlun 是否应实现正式、TP-aware 的 vocab 分块 fallback，以及对应 padding/gather 合同。

相关问题：#362、#366、#382 都涉及输出正确性或 Qwen3.6 部署，但本 Issue 提供的是 unquantized lm-head 的独立 `F.linear` 最小数值复现。

**Comment by CarrotSwordsman** [comment]
在较新的软件栈上尝试复现，**没有复现**，共 36 组配置全部与 CPU fp32 reference 对齐。数据可能对判断此问题是否已被修复有帮助。

## 环境对比

| 项目 | 你的环境 | 我的环境 |
| --- | --- | --- |
| XPU | P800 PCIe 96 GiB | **P800 OAM 96 GiB** |
| Driver | `5.0.21.26` | **`515.58`** |
| XPU-RT | `5.0.21` | **`10.2`** |
| Torch | `2.5.1+cu118` | **`2.9.0+cu129`** |
| 集成方式 | vllm-kunlun `0.15.1.dev0` | 另一套 XPU torch 构建（`torch_xmlir` + `SYMBOL_REWRITE`） |

驱动、XPU-RT 和 torch 三者跨度都很大，所以下面的结论只说明"在这一版栈上该行为不存在"，不能直接推翻你的观测。

## 测试方式

由于问题定位在 `torch.nn.functional.linear` 这个公开边界上，我用**合成权重**做了 model-free 的复现，不依赖 Qwen3.6-27B 权重，因此结论与具体 checkpoint 无关。每组都比较三者：单次完整 `F.linear`、按 vocab 行分块后 `torch.cat`、以及 CPU fp32 分块 reference。

**1. 主 shape（你报告的 `M=1, N=248320, K=5120`, fp16）**

```
full F.linear        nan=0 inf=0 max_abs=0.00369 rel_l2=3.065e-4 top1=1429
chunked rows=131072  nan=0 inf=0 max_abs=0.00369 rel_l2=3.065e-4 top1=1429
chunked rows=65536   ... 同上
chunked rows=32768   ... 同上
chunked rows=4096    ... 同上
```

完整与分块结果**逐位相同**，`rel_l2 ≈ 3e-4` 属于 fp16 相对 fp32 的正常误差。权重 H2D 传输也校验为 bit-exact。

**2. 参数扫描（13 组，全部 OK）**

- `N` = 32768 / 65536 / 131072 / 151936 / 248320 / 262144 — 跨过你报告的 vocab 尺寸，未见阈值
- `M` = 1 / 2 / 8 / 32
- 权重 std = 0.005 / 0.02 / 0.05 / 0.2
- bf16 对照（`rel_l2 ≈ 1.65e-3`，符合 bf16 精度）

**3. 边界场景（10 组，全部 OK）** — 这些是合成高斯权重覆盖不到、但真实 lm-head 可能具备的特征：

- 离群行（×50 放大）、1000 个全零行、含接近 fp16 上限的行（60.0）
- 非对齐 `N` = 248321 / 248319 / 248327 / 250001
- 非连续权重：大 buffer 切片、以及 `is_contiguous=False` 的转置视图
- 连续 30 次调用，逐次结果完全一致（排除多次 launch 后的状态污染）

## 一个可能相关的发现

`FORCE_NN_LINEAR=1` 确实会切到**另一个 kernel**：同样的输入下 `rel_l2` 从 `3.065e-4` 稳定变为 `2.069e-4`（精度更高）。在我的环境里两条路径都正确，但这说明 `F.linear` 在 XPU 上至少有两条实现路径。如果你的环境里两者行为不同，对比一下可能有助于定位。

## 复现脚本

不依赖模型权重，可直接在你的镜像里跑：

```python
import torch, torch.nn.functional as F

M, N, K = 1, 248320, 5120
torch.manual_seed(0)
w_cpu = (torch.randn(N, K) * 0.02).half().contiguous()
h_cpu = torch.randn(M, K).half().contiguous()

# CPU fp32 reference，分块只为控制内存
ref = torch.cat([F.linear(h_cpu.float(), w_cpu[i:i+8192].float())
                 for i in range(0, N, 8192)], dim=-1)

w, h = w_cpu.cuda(), h_cpu.cuda()
torch.cuda.synchronize()
full = F.linear(h, w).float().cpu()
chunked = torch.cat([F.linear(h, w[i:i+32768]) for i in range(0, N, 32768)],
                    dim=-1).float().cpu()
torch.cuda.synchronize()

n = ref.norm()
print("rel_full   ", ((full - ref).norm() / n).item())
print("rel_chunked", ((chunked - ref).norm() / n).item())
print("paths bit-equal", torch.equal(full, chunked))
print("top1", full.argmax().item(), chunked.argmax().item(), ref.argmax().item())
```

在我的环境输出 `rel_full ≈ rel_chunked ≈ 3.07e-4`、`paths bit-equal True`、三个 top1 相同。

## 需要说明的局限

- 用的是合成权重。虽然额外覆盖了离群值、零行、近上限值和非连续布局，仍不等同于真实 Qwen3.6-27B lm-head 的数值分布。
- 硬件是 OAM 版而非 PCIe 版。
- 我这套 XPU torch 与 `vllm-kunlun 0.15.1` 的运行时不是同一套，尽管脚本本身不启动 vLLM。

## 建议

如果你还能访问原环境，跑一下上面这段脚本会很有价值：

- 若在你的环境**能**复现 → 说明与真实权重无关，是这一版驱动/XPU-RT/torch 的问题，且新版本已修复；
- 若在你的环境**不能**复现 → 说明触发条件依赖真实 lm-head 权重的数值分布，那么合成权重的排查方向需要调整。

无论哪种结果，都能把范围收窄一大截。我这边有 P800 可以继续配合验证。

**Comment by AlanJager** [comment]
## v0.25.1 复测结果：未复现，关闭 Issue

2026-09-21 在原问题同类硬件和驱动环境上，使用新镜像重新执行了 model-free 与真实权重两组验证。

### 环境

- XPU：P800 PCIe 96 GiB
- OS / kernel：Ubuntu 22.04 / Linux 5.15.0-161-generic
- Driver / XPU-RT：`5.0.21.26` / `5.0.21`，与原报告一致
- 镜像：`vllm-kunlun:p800-aios-v0.25.1-slim`
- Image ID：`sha256:9c7522884d0c22e3a83e9785db508852a0f791bf3ba2b055be004cf4b48837a1`
- vLLM / Torch：`0.25.1` / `2.9.0+cu128`

### Model-free 结果

使用原问题 shape `M=1, N=248320, K=5120`、FP16，比较完整 `F.linear`、按 65536 行分块的 `F.linear` 和 CPU FP32 reference：

- NaN / Inf：`0 / 0`
- relative L2：`3.058e-4`
- 完整与分块结果：逐位一致
- 完整、分块、CPU reference 的 top-1：一致

### 真实 Qwen3.6-27B 权重结果

通过只读 JuiceFS 挂载读取原模型完整权重，确认 `lm_head.weight=[248320,5120]`。按原 Issue 的三个 seed 重测：

| seed | NaN / Inf | relative L2 | 完整与分块 | top-1 |
| --- | ---: | ---: | --- | --- |
| `20260714` | `0 / 0` | `3.011e-4` | 逐位一致 | 与 CPU 一致 |
| `20260715` | `0 / 0` | `3.013e-4` | 逐位一致 | 与 CPU 一致 |
| `20260716` | `0 / 0` | `3.060e-4` | 逐位一致 | 与 CPU 一致 |

原环境中完整 GEMM 会固定出现 `1178 NaN + 2 Inf`、relative L2 大幅异常且 top-1 错误；在 v0.25.1 新镜像中，上述异常均未出现，完整 GEMM 与分块 GEMM 逐位相同。

### 结论

在本次验证的 P800 PCIe、TP1、FP16、unquantized、无 bias 场景中，v0.25.1 已无法复现该问题，可以认为当前镜像使用的新软件栈已经解决原始故障。因此关闭本 Issue。

边界说明：本次未覆盖 TP2、长上下文、并发压力和其他 dtype。
